pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/TransformExpressionOperator.java

```diff
@@ -0,0 +1,161 @@
+    @Override
+    protected int[] initialValue() {
+      return new int[DocIdSetPlanNode.MAX_DOC_PER_CALL];
+    }
+  };
+
+  /**
+   * Constructor for the class
+   *
+   * @param indexSegment Segment to process
+   * @param projectionOperator Projection operator
+   * @param expressionTree Expression tree to evaluate
+   */
+  public TransformExpressionOperator(IndexSegment indexSegment, MProjectionOperator projectionOperator,
+      TransformExpressionTree expressionTree) {
+
+    Preconditions.checkArgument((indexSegment != null));
+    Preconditions.checkArgument((projectionOperator != null));
+    Preconditions.checkArgument((expressionTree != null));
+
+    _indexSegment = indexSegment;
+    _projectionOperator = projectionOperator;
+
+    _expressionEvaluator = new DefaultExpressionEvaluator(indexSegment, expressionTree);
+  }
+
+  @Override
+  public Block getNextBlock() {
+    return getNextBlock(new BlockId(_nextBlockCallCounter++));
+  }
+
+  @Override
+  public Block getNextBlock(BlockId blockId) {
+    // Only one block supported, currently.
+    if (blockId.getId() > 0) {
+      return null;
+    }
+
+    int numDocsScanned = 0;
+    final long startTimeMillis = System.currentTimeMillis();
+
+    _expressionEvaluator.init();
+    while (_projectionOperator.nextBlock() != null) {
+      ProjectionBlock currentBlock = _projectionOperator.getCurrentBlock();
+      Block docIdSetBlock = currentBlock.getDocIdSetBlock();
+      numDocsScanned = processBlock(currentBlock, docIdSetBlock, numDocsScanned);
+    }
+    _expressionEvaluator.finish();
+
+    IntermediateResultsBlock result = new IntermediateResultsBlock(_expressionEvaluator.getResult());
+    result.setTotalRawDocs(_indexSegment.getSegmentMetadata().getTotalRawDocs());
+    result.setTimeUsedMs(System.currentTimeMillis() - startTimeMillis);
+
+    return result;
+  }
+
+  @Override
+  public String getOperatorName() {
+    return OPERATOR_NAME;
+  }
+
+  @Override
+  public boolean open() {
+    return _projectionOperator.open();
+  }
+
+  @Override
+  public boolean close() {
+    return _projectionOperator.close();
+  }
+
+  /**
+   * Process a block of docIdSets by evaluating the transform expression on each doc.
+   *
+   * @param projectionBlock Project block to process
+   * @param block DocIdSet block to process
+   * @param numDocsScanned Number of documents already scanned
+   * @return Number of documents scanned (including the original passed-in value)
+   */
+  private int processBlock(ProjectionBlock projectionBlock, Block block, int numDocsScanned) {
+
+    if (block instanceof DocIdSetBlock) {
+      DocIdSetBlock docIdSetBlock = (DocIdSetBlock) block;
+      int length = docIdSetBlock.getSearchableLength();
+
+      _expressionEvaluator.evaluate(docIdSetBlock.getDocIdSet(), length);
+      numDocsScanned += length;
+
+    } else {
+
+      BlockDocIdSet blockDocIdSet = projectionBlock.getBlockDocIdSet();
```

Comment:
For ProjectionBlock, we have both getDocIdSetBlock() and getBlockDocIdSet()? LOL